compiler/rustc_serialize/src/leb128.rs

```diff
@@ -51,28 +51,35 @@ macro_rules! impl_read_unsigned_leb128 {
     ($fn_name:ident, $int_ty:ty) => {
         #[inline]
         pub fn $fn_name(slice: &[u8], position: &mut usize) -> $int_ty {
-            // The first iteration of this loop is unpeeled. This is a
-            // performance win because this code is hot and integer values less
-            // than 128 are very common, typically occurring 50-80% or more of
-            // the time, even for u64 and u128.
-            let byte = slice[*position];
-            *position += 1;
-            if (byte & 0x80) == 0 {
-                return byte as $int_ty;
-            }
-            let mut result = (byte & 0x7F) as $int_ty;
-            let mut shift = 7;
-            loop {
-                let byte = slice[*position];
-                *position += 1;
+            #[inline]
+            fn inner(slice: &[u8], position: &mut usize) -> Option<$int_ty> {
```

Comment:
Why the `inner`? Couldn't you just use indexing syntax for the panicking gets? Do you want to minimize the number of panic paths to minimize code size?

Reply:
Yes. The code that I'm porting this from is a must-never-panic parser, so these failures are hoisted to an outer layer with `?`. Then in picking through the existing codegen in rustc, I noticed that we have two separate calls to `core::panicking::panic_bounds_check` at the bottom of this function, because the use of panicking indexing requires rustc to preserve the line numbers in the panic message. But nobody cares exactly which line panicked, only that leb128 decoding failed.

I'm pushing a commit to establish if this matters.

Reply:
Well, looks like it matters 🙃